File: crates/holos-tda-py/python/holos_tda/bipersistence.py
"""Finite degree-Rips bipersistence with checked module evidence.

The module stores the complete finite H1 diagram on the degree-Rips grid.
Queries return dictionaries with explicit grade, map, class, and coordinate
fields. The underlying ``HOLOSBP`` bytes can be saved as a reproducible claim.
"""

from . import _core
from ._coerce import _triplets

__all__ = [
    "Bipersistence",
    "BipersistenceArtifact",
    "degree_rips_bipersistence",
    "load_bipersistence_artifact",
]


def _term_records(terms):
    return [
        {"basis_index": int(index), "coefficient": int(coefficient)}
        for index, coefficient in terms
    ]


def _map_record(raw):
    lower, upper, rank, columns = raw
    return {
        "lower": tuple(lower),
        "upper": tuple(upper),
        "rank": int(rank),
        "columns": [
            {
                "source_basis_index": int(source),
                "image": _term_records(image),
            }
            for source, image in columns
        ],
    }


def _atlas_record(raw):
    base, base_class, extensions, regions = raw
    return {
        "base": tuple(base),
        "base_class": _term_records(base_class),
        "extensions": [
            {
                "grade": tuple(grade),
                "kind": kind,
                "class": _term_records(class_terms),
                "ambiguity": [_term_records(direction) for direction in ambiguity],
            }
            for grade, kind, class_terms, ambiguity in extensions
        ],
        "regions": [
            {
                "region_index": int(index),
                "kind": kind,
                "ambiguity_rank": int(ambiguity_rank),
                "grades": [tuple(grade) for grade in grades],
            }
            for index, kind, ambiguity_rank, grades in regions
        ],
    }


def _coordinate_record(raw):
    (
        space,
        modulus,
        scale,
        field_multiplier,
        class_terms,
        source,
        integral,
        divisibility,
        potential,
        phase,
        diagnostics,
    ) = raw
    energy, max_residual, relative_residual, iterations, tolerance = diagnostics
    return {
        "space": space,
        "modulus": int(modulus),
        "scale": float(scale),
        "field_multiplier": int(field_multiplier),
        "class": _term_records(class_terms),
        "source": [
            {"simplex": list(simplex), "coefficient": int(coefficient)}
            for simplex, coefficient in source
        ],
        "integral": [
            {"u": int(u), "v": int(v), "coefficient": int(coefficient)}
            for u, v, coefficient in integral
        ],
        "divisibility": int(divisibility),
        "potential": [float(value) for value in potential],
        "phase": [float(value) for value in phase],
        "energy": float(energy),
        "max_residual": float(max_residual),
        "relative_residual": float(relative_residual),
        "iterations": int(iterations),
        "tolerance": float(tolerance),
    }


def _family_record(raw):
    base, base_class, entries = raw
    return {
        "base": tuple(base),
        "base_class": _term_records(base_class),
        "entries": [
            {
                "grade": tuple(grade),
                "kind": kind,
                "coordinate": (
                    None if coordinate is None else _coordinate_record(coordinate)
                ),
            }
            for grade, kind, coordinate in entries
        ],
    }


def _summary(raw):
    (
        vertices,
        edges,
        nodes,
        cover_maps,
        rectangles,
        regions,
        class_atlases,
        circular_families,
    ) = raw
    return {
        "vertices": int(vertices),
        "edges": int(edges),
        "nodes": int(nodes),
        "cover_maps": int(cover_maps),
        "rectangles": int(rectangles),
        "regions": int(regions),
        "class_atlases": int(class_atlases),
        "circular_families": int(circular_families),
    }


class Bipersistence:
    """A checked finite degree-Rips H1 module.

    Construct an instance with :func:`degree_rips_bipersistence`. Grades are
    ``(scale_index, density_index)`` pairs. The density axis addresses the
    descending ``minimum_degrees`` property.
    """

    def __init__(self, inner):
        self._inner = inner

    @property
    def modulus(self):
        """Return the prime coefficient modulus."""
        return int(self._inner.modulus)

    @property
    def scales(self):
        """Return scale values in ascending order."""
        return [float(value) for value in self._inner.scales]

    @property
    def minimum_degrees(self):
        """Return minimum-degree levels in descending order."""
        return [int(value) for value in self._inner.minimum_degrees]

    @property
    def node_ranks(self):
        """Return one record for each grid node."""
        return [
            {
                "grade": (int(scale), int(density)),
                "rank": int(rank),
            }
            for scale, density, rank in self._inner.node_ranks
        ]

    @property
    def cover_maps(self):
        """Return all checked horizontal and vertical cover maps."""
        return [_map_record(value) for value in self._inner.cover_maps]

    def map_rank(self, lower, upper):
        """Return the exact map rank between comparable grades."""
        return int(self._inner.map_rank(tuple(lower), tuple(upper)))

    def map(self, lower, upper):
        """Return a canonical matrix for the map between two grades."""
        return _map_record(self._inner.map(tuple(lower), tuple(upper)))

    def rectangle_rank(self, lower, upper):
        """Return the generalized rank of a closed parameter rectangle."""
        return int(self._inner.rectangle_rank(tuple(lower), tuple(upper)))

    def region_rank(self, grades):
        """Return the generalized rank of a connected finite region."""
        return int(self._inner.region_rank([tuple(grade) for grade in grades]))

    def class_atlas(self, base, class_):
        """Return the exact extension atlas of one nonzero class."""
        return _atlas_record(self._inner.class_atlas(tuple(base), _normalize_terms(class_)))

    def circular_family(
        self, base, class_, tolerance=1e-10, max_iterations=10_000
    ):
        """Return checked circular coordinates for unique class extensions."""
        return _family_record(
            self._inner.circular_family(
                tuple(base),
                _normalize_terms(class_),
                float(tolerance),
                int(max_iterations),
            )
        )

    def record_rectangle(self, lower, upper):
        """Add one checked rectangle-rank claim to the artifact."""
        self._inner.record_rectangle(tuple(lower), tuple(upper))

    def record_region(self, grades):
        """Add one checked connected-region rank claim to the artifact."""
        self._inner.record_region([tuple(grade) for grade in grades])

    def record_class_atlas(self, base, class_):
        """Add one checked class atlas to the artifact."""
        return _atlas_record(
            self._inner.record_class_atlas(tuple(base), _normalize_terms(class_))
        )

    def record_circular_family(
        self, base, class_, tolerance=1e-10, max_iterations=10_000
    ):
        """Add checked circular data for a stored class atlas."""
        return _family_record(
            self._inner.record_circular_family(
                tuple(base),
                _normalize_terms(class_),
                float(tolerance),
                int(max_iterations),
            )
        )

    @property
    def artifact(self):
        """Return canonical ``HOLOSBP`` bytes for the current claims."""
        return bytes(self._inner.artifact)

    @property
    def artifact_summary(self):
        """Return structural counts for the current artifact."""
        return _summary(self._inner.artifact_summary)


class BipersistenceArtifact:
    """A decoded, self-contained, and verified ``HOLOSBP`` artifact."""

    def __init__(self, inner):
        self._inner = inner

    @property
    def bytes(self):
        """Return canonical artifact bytes."""
        return bytes(self._inner.bytes)

    @property
    def summary(self):
        """Return structural counts from the artifact."""
        return _summary(self._inner.summary)

    @property
    def rectangles(self):
        """Return stored generalized rectangle-rank claims."""
        return [
            {
                "lower": tuple(lower),
                "upper": tuple(upper),
                "rank": int(rank),
            }
            for lower, upper, rank in self._inner.rectangles
        ]

    @property
    def regions(self):
        """Return stored generalized connected-region rank claims."""
        return [
            {
                "grades": [tuple(grade) for grade in grades],
                "rank": int(rank),
            }
            for grades, rank in self._inner.regions
        ]

    @property
    def class_atlases(self):
        """Return stored class-extension atlases."""
        return [_atlas_record(value) for value in self._inner.class_atlases]

    def verify(self):
        """Replay and verify the exact artifact claims."""
        self._inner.verify()


def _normalize_terms(class_):
    return [(int(index), int(coefficient)) for index, coefficient in class_]


def degree_rips_bipersistence(
    n,
    triplets,
    threshold=None,
    modulus=47,
    *,
    scales=None,
    minimum_degrees=None,
):
    """Build a checked degree-Rips H1 module from sparse weighted edges.

    ``triplets`` contains ``(u, v, weight)`` rows. By default, the module uses
    every induced critical scale and minimum degree. Pass both ``scales`` and
    ``minimum_degrees`` to declare a smaller exact grid. The scale values must
    increase, and the minimum degrees must decrease and end at zero.
    """
    inner = _core.degree_rips_bipersistence(
        int(n),
        _triplets(triplets),
        None if threshold is None else float(threshold),
        int(modulus),
        scales=(None if scales is None else [float(value) for value in scales]),
        minimum_degrees=(
            None
            if minimum_degrees is None
            else [int(value) for value in minimum_degrees]
        ),
    )
    return Bipersistence(inner)


def load_bipersistence_artifact(data):
    """Decode and verify canonical ``HOLOSBP`` bytes."""
    return BipersistenceArtifact(_core.load_bipersistence_artifact(bytes(data)))

File: crates/holos-tda/src/cli/bipersistence.rs
use std::fmt::Write as _;
use std::path::{Path, PathBuf};

use clap::Parser;

use super::{InputFormat, read_proof_input, version_string, write_via_temporary};
use crate::{
    BipersistenceArtifact, BipersistenceArtifactLimits, BipersistenceModule,
    BipersistenceRectangleClaim, BipersistenceRegionClaim, BipersistenceTerm,
    CircularCoordinateFamily, CircularCoordinateParams, ClassExtensionKind, CohomologyClassAtlas,
    CohomologyLimits, Result,
};

mod input;

#[derive(Parser)]
#[command(
    name = "holos bipersistence",
    version = version_string(),
    about = "Build and certify finite degree-Rips H1 bipersistence",
    after_help = "Grid coordinates are zero-based indices. A class cocycle contains `u v coefficient` rows. Check OUTPUT with: holos-check OUTPUT"
)]
pub(super) struct BipersistenceCli {
    /// Input point cloud, lower-distance matrix, or sparse weighted graph
    input: PathBuf,

    /// Output `HOLOSBP` artifact
    output: PathBuf,

    /// Input format. Inferred when omitted. Sparse is never inferred
    #[arg(long, value_enum)]
    format: Option<InputFormat>,

    /// Largest degree-Rips scale. The input default is used when omitted
    #[arg(long, value_name = "T")]
    threshold: Option<f64>,

    /// Scale on a declared finite grid. Repeat in strict ascending order
    #[arg(long = "scale", value_name = "T")]
    scales: Vec<f64>,

    /// Minimum degree on a declared finite grid. Repeat in strict descending order
    #[arg(long = "minimum-degree", value_name = "K")]
    minimum_degrees: Vec<usize>,

    /// Coefficient field Z/p; must be a prime below 32768
    #[arg(long, value_name = "P", default_value_t = 47)]
    modulus: u32,

    /// Rectangle corners as `LOW_SCALE LOW_DENSITY HIGH_SCALE HIGH_DENSITY`
    #[arg(long = "rectangle", value_names = ["S0", "D0", "S1", "D1"], num_args = 4)]
    rectangles: Vec<usize>,

    /// File with `SCALE DENSITY` rows for one connected region. Repeat as needed
    #[arg(long = "region", value_name = "FILE")]
    regions: Vec<PathBuf>,

    /// Canonical basis class as `SCALE DENSITY BASIS`. Repeat as needed
    #[arg(long = "class", value_names = ["S", "D", "B"], num_args = 3)]
    classes: Vec<usize>,

    /// Cocycle class as `SCALE DENSITY FILE`. Repeat as needed
    #[arg(long = "class-cocycle", value_names = ["S", "D", "FILE"], num_args = 3)]
    class_cocycles: Vec<String>,

    /// Add a checked circular coordinate at every unique class extension
    #[arg(long)]
    circular: bool,

    /// Largest accepted relative normal-equation residual
    #[arg(long, value_name = "R", default_value_t = 1e-10)]
    tolerance: f64,

    /// Largest harmonic solver iteration count
    #[arg(long, value_name = "N", default_value_t = 10_000)]
    max_iterations: usize,

    /// Write a JSON report with axes, ranks, fibers, and optional phases
    #[arg(long, value_name = "FILE")]
    report: Option<PathBuf>,

    /// Worker budget for input parsing
    #[arg(long, value_name = "N", default_value_t = 1)]
    threads: usize,

    /// Largest accepted output artifact
    #[arg(long, value_name = "BYTES", default_value_t = 1usize << 30)]
    max_artifact_bytes: usize,
}

pub(super) fn run(cli: BipersistenceCli) -> Result<()> {
    let input_threshold = input::input_threshold(&cli);
    let graph = read_proof_input(&cli.input, cli.format, cli.threads, input_threshold)?;
    let degree_rips = input::build_degree_rips(&graph, &cli, input_threshold)?;
    let limits = input::artifact_limits(&cli);
    let (mut artifact, module) = BipersistenceArtifact::build(&degree_rips, cli.modulus, limits)?;
    input::record_rectangles(&mut artifact, &module, &cli.rectangles, limits)?;
    input::record_regions(
        &mut artifact,
        &module,
        &cli.regions,
        cli.max_artifact_bytes,
        limits,
    )?;
    let selections = input::selected_classes(&module, &cli)?;
    let circular_params = CircularCoordinateParams {
        tolerance: cli.tolerance,
        max_iterations: cli.max_iterations,
        cohomology: CohomologyLimits::default(),
    };
    let (atlases, families) = input::record_class_claims(
        &mut artifact,
        &module,
        selections,
        cli.circular,
        circular_params,
        limits,
    )?;

    if let Some(path) = &cli.report {
        write_report(
            path,
            &module,
            artifact.rectangles(),
            artifact.regions(),
            &atlases,
            &families,
        )?;
    }
    write_artifact(&cli.output, &artifact, limits)
}

fn write_artifact(
    output: &Path,
    artifact: &BipersistenceArtifact,
    limits: BipersistenceArtifactLimits,
) -> Result<()> {
    let bytes = artifact.encode(limits)?;
    write_via_temporary(output, &bytes)?;
    let summary = artifact.summary();
    println!(
        "wrote degree-Rips H1 module with {} nodes, {} cover maps, {} rectangle claims, {} connected-region claims, {} class atlases, and {} circular families to {}",
        summary.nodes,
        summary.cover_maps,
        summary.rectangles,
        summary.regions,
        summary.class_atlases,
        summary.circular_families,
        output.display(),
    );
    Ok(())
}

fn write_report(
    path: &Path,
    module: &BipersistenceModule,
    rectangles: &[BipersistenceRectangleClaim],
    regions: &[BipersistenceRegionClaim],
    atlases: &[CohomologyClassAtlas],
    families: &[CircularCoordinateFamily],
) -> Result<()> {
    let mut json = String::new();
    write_report_header(&mut json, module);
    write_report_nodes(&mut json, module);
    write_report_rectangles(&mut json, rectangles);
    write_report_regions(&mut json, regions);
    write_report_atlases(&mut json, atlases);
    write_report_families(&mut json, families);
    write_via_temporary(path, json.as_bytes())
}

fn write_report_header(json: &mut String, module: &BipersistenceModule) {
    json.push_str("{\n  \"format\": \"holos-degree-rips-report-v1\",\n  \"modulus\": ");
    write!(json, "{},\n  \"scales\": [", module.modulus())
        .expect("writing to a string cannot fail");
    for (position, scale) in module.scales().iter().enumerate() {
        if position != 0 {
            json.push(',');
        }
        write!(json, "{scale}").expect("writing to a string cannot fail");
    }
    json.push_str("],\n  \"minimum_degrees\": [");
    for (position, degree) in module.minimum_degrees().iter().enumerate() {
        if position != 0 {
            json.push(',');
        }
        write!(json, "{degree}").expect("writing to a string cannot fail");
    }
    json.push_str("],\n  \"nodes\": [");
}

fn write_report_nodes(json: &mut String, module: &BipersistenceModule) {
    for (position, node) in module.nodes().iter().enumerate() {
        if position != 0 {
            json.push(',');
        }
        write!(
            json,
            "\n    {{\"scale_index\":{},\"density_index\":{},\"rank\":{}}}",
            node.grade.scale(),
            node.grade.density(),
            node.rank,
        )
        .expect("writing to a string cannot fail");
    }
    json.push_str("\n  ],\n  \"rectangles\": [");
}

fn write_report_rectangles(json: &mut String, rectangles: &[BipersistenceRectangleClaim]) {
    for (position, claim) in rectangles.iter().enumerate() {
        if position != 0 {
            json.push(',');
        }
        let low = claim.rectangle.lower;
        let high = claim.rectangle.upper;
        write!(
            json,
            "\n    {{\"low\":[{},{}],\"high\":[{},{}],\"rank\":{}}}",
            low.scale(),
            low.density(),
            high.scale(),
            high.density(),
            claim.rank,
        )
        .expect("writing to a string cannot fail");
    }
    json.push_str("\n  ],\n  \"regions\": [");
}

fn write_report_regions(json: &mut String, regions: &[BipersistenceRegionClaim]) {
    for (position, claim) in regions.iter().enumerate() {
        if position != 0 {
            json.push(',');
        }
        json.push_str("\n    {\"grades\":[");
        for (grade_position, grade) in claim.region.grades().iter().enumerate() {
            if grade_position != 0 {
                json.push(',');
            }
            write!(json, "[{},{}]", grade.scale(), grade.density())
                .expect("writing to a string cannot fail");
        }
        write!(json, "],\"rank\":{}}}", claim.rank).expect("writing to a string cannot fail");
    }
    json.push_str("\n  ],\n  \"class_atlases\": [");
}

fn write_report_atlases(json: &mut String, atlases: &[CohomologyClassAtlas]) {
    for (position, atlas) in atlases.iter().enumerate() {
        if position != 0 {
            json.push(',');
        }
        write_atlas_json(json, atlas);
    }
    json.push_str("\n  ],\n  \"circular_families\": [");
}

fn write_report_families(json: &mut String, families: &[CircularCoordinateFamily]) {
    for (position, family) in families.iter().enumerate() {
        if position != 0 {
            json.push(',');
        }
        write_family_json(json, family);
    }
    json.push_str("\n  ]\n}\n");
}

fn write_atlas_json(json: &mut String, atlas: &CohomologyClassAtlas) {
    write!(
        json,
        "\n    {{\"base\":[{},{}],\"class\":[",
        atlas.base_grade.scale(),
        atlas.base_grade.density(),
    )
    .expect("writing to a string cannot fail");
    write_terms_json(json, &atlas.base_class);
    json.push_str("],\"extensions\":[");
    for (position, extension) in atlas.extensions.iter().enumerate() {
        if position != 0 {
            json.push(',');
        }
        write!(
            json,
            "{{\"grade\":[{},{}],\"kind\":\"{}\",\"ambiguity_rank\":{}}}",
            extension.grade.scale(),
            extension.grade.density(),
            extension_kind(extension.kind),
            extension.ambiguity.len(),
        )
        .expect("writing to a string cannot fail");
    }
    write!(json, "],\"region_count\":{}}}", atlas.regions.len())
        .expect("writing to a string cannot fail");
}

fn write_family_json(json: &mut String, family: &CircularCoordinateFamily) {
    write!(
        json,
        "\n    {{\"base\":[{},{}],\"entries\":[",
        family.base_grade.scale(),
        family.base_grade.density(),
    )
    .expect("writing to a string cannot fail");
    for (position, entry) in family.entries.iter().enumerate() {
        if position != 0 {
            json.push(',');
        }
        write!(
            json,
            "{{\"grade\":[{},{}],\"kind\":\"{}\",\"phase\":",
            entry.grade.scale(),
            entry.grade.density(),
            extension_kind(entry.extension),
        )
        .expect("writing to a string cannot fail");
        match &entry.coordinate {
            None => json.push_str("null"),
            Some(coordinate) => {
                json.push('[');
                for (phase_position, phase) in coordinate.phase.iter().enumerate() {
                    if phase_position != 0 {
                        json.push(',');
                    }
                    write!(json, "{phase}").expect("writing to a string cannot fail");
                }
                json.push(']');
            }
        }
        json.push('}');
    }
    json.push_str("]}");
}

fn write_terms_json(json: &mut String, terms: &[BipersistenceTerm]) {
    for (position, term) in terms.iter().enumerate() {
        if position != 0 {
            json.push(',');
        }
        write!(json, "[{},{}]", term.basis_index, term.coefficient)
            .expect("writing to a string cannot fail");
    }
}

fn extension_kind(kind: ClassExtensionKind) -> &'static str {
    match kind {
        ClassExtensionKind::Unique => "unique",
        ClassExtensionKind::Ambiguous => "ambiguous",
        ClassExtensionKind::NoExtension => "no_extension",
    }
}
